src/utils.js

```diff
@@ -40,3 +42,74 @@ Gio._promisify(Gio.Subprocess.prototype, 'communicate_utf8_async',
     'communicate_utf8_finish');
 Gio._promisify(Gio.Subprocess.prototype, 'wait_check_async',
     'wait_check_finish');
+
+function _appInfoFromKeyFile(keyfile) {
+    const startGroup = keyfile.get_start_group();
+    if (!startGroup || startGroup !== GLib.KEY_FILE_DESKTOP_GROUP)
+        throw new Error('Invalid .desktop file');
+
+    const type = keyfile.get_string(GLib.KEY_FILE_DESKTOP_GROUP,
+        GLib.KEY_FILE_DESKTOP_KEY_TYPE);
+    if (type !== GLib.KEY_FILE_DESKTOP_TYPE_APPLICATION)
+        throw new Error('Not an application .desktop file');
+
+    const name = keyfile.get_locale_string(GLib.KEY_FILE_DESKTOP_GROUP,
+        GLib.KEY_FILE_DESKTOP_KEY_NAME, null);
+
+    let icon = null;
+    try {
+        let iconName = keyfile.get_locale_string(GLib.KEY_FILE_DESKTOP_GROUP,
+            GLib.KEY_FILE_DESKTOP_KEY_ICON, null);
+        if (GLib.path_is_absolute(iconName)) {
+            const file = Gio.File.new_for_path(iconName);
+            icon = new Gio.FileIcon({file});
+        } else {
+            // Work around a common mistake in desktop files
+            if (iconName.endsWith('.png') || iconName.endsWith('.xpm') ||
+                iconName.endsWith('.xvg'))
+                iconName = iconName.slice(0, -4);
+            icon = new Gio.ThemedIcon({name: iconName});
+        }
+    } catch (e) {
+        void e;
+    }
+
+    return {name, icon};
+}
+
+
+function appInfoForAppId(id) {
+    // Check GDesktopAppInfo's default search paths first
+    const appInfo = Gio.DesktopAppInfo.new(`${id}.desktop`);
+    if (appInfo) {
+        return {
+            name: appInfo.get_name(),
+            icon: appInfo.get_icon(),
+        };
+    }
+
+    // Try to load manually from the .desktop key file
```

Comment:
FWIW the "canonical" hack to work around GDesktopAppInfo tripping on the flatpak `Exec` line is to override it to e.g. `/bin/true`. See e.g. https://github.com/endlessm/eos-discovery-feed/blob/a8552da04ea5935698ca3feeac904c38dc200b70/src/main.js#L122